### PR TITLE
Cancel order with IPN 

### PR DESCRIPTION
**Pull request by Francois-Gomis**

### CODE REVIEW

File: Test/Unit/Gateway/Request/PaymentDataBuilderTest.php [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            'locale' => self::LOCALE,
+            'expires_after' => 0,
+            'custom_data' => [
+                'order_id' => self::INCREMENT_ID,
+                'quote_id' => self::QUOTE_ID,
+            ]
+        ];
+
+        return [
+            'Non trigger Payload' => [
```
Comment (PaulRouss3l) [inline Test/Unit/Gateway/Request/PaymentDataBuilderTest.php:101]:
```suggestion
            'dont trigger Payload' => [
```

File: Test/Unit/Gateway/Request/PaymentDataBuilderTest.php [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                'order' => [
+                    'is_trigger' => false,
+                    'plan_has_trigger' => false,
+                ],
+                'final_payload' => [
+                    'payment' => $paymentBase
+                ]
+            ],
+            'Trigger is enable but plan is not trigger' => [
```
Comment (PaulRouss3l) [inline Test/Unit/Gateway/Request/PaymentDataBuilderTest.php:110]:
```suggestion
            'Trigger is enable but plan is not trigged' => [
```

File: Test/Unit/Gateway/Request/PaymentDataBuilderTest.php [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+        $this->config->expects($this->once())->method('getPaymentPlansConfig')->willReturn($paymentPlansConfigMock);
+        $this->config->expects($this->once())->method('getReturnUrl')->willReturn(self::RETURN_URL);
+        $this->config->expects($this->once())->method('getIpnCallbackUrl')->willReturn(self::IPN_URL);
+        $this->config->expects($this->once())->method('getCustomerCancelUrl')->willReturn(self::CANCEL_URL);
+        $this->config->expects($this->once())->method('getFailureReturnUrl')->willReturn(self::FAILURE_URL);
+
+        $this->locale->expects($this->once())->method('getLocale')->willReturn(self::LOCALE);
+
+        $this->configHelper->expects($this->once())->method('triggerIsEnabled')->willReturn($order['is_trigger']);
```
Comment (PaulRouss3l) [inline Test/Unit/Gateway/Request/PaymentDataBuilderTest.php:74]:
```suggestion
        $this->configHelper->expects($this->once())->method('triggerIsEnabled')->willReturn($order['is_triggered']);
```
ou
```suggestion
        $this->configHelper->expects($this->once())->method('triggerIsEnabled')->willReturn($order['triggered']);
```

File: Test/Unit/Gateway/Request/PaymentDataBuilderTest.php [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            'custom_data' => [
+                'order_id' => self::INCREMENT_ID,
+                'quote_id' => self::QUOTE_ID,
+            ]
+        ];
+
+        return [
+            'Non trigger Payload' => [
+                'order' => [
+                    'is_trigger' => false,
```
Comment (PaulRouss3l) [inline Test/Unit/Gateway/Request/PaymentDataBuilderTest.php:103]: is_triggered ou triggered

File: Test/Unit/Gateway/Request/PaymentDataBuilderTest.php [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+                    'plan_has_trigger' => false,
+                ],
+                'final_payload' => [
+                    'payment' => $paymentBase
+                ]
+            ],
+            'Trigger is enable but plan is not trigger' => [
+                'order' => [
+                    'is_trigger' => true,
```
Comment (PaulRouss3l) [inline Test/Unit/Gateway/Request/PaymentDataBuilderTest.php:112]: is_triggered ou triggered

File: Test/Unit/Gateway/Request/PaymentDataBuilderTest.php [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                    'plan_has_trigger' => false,
+                ],
+                'final_payload' => [
+                    'payment' => $paymentBase
+                ]
+            ],
+            'Trigger is enable and plan is trigger' => [
+                'order' => [
+                    'is_trigger' => true,
```
Comment (PaulRouss3l) [inline Test/Unit/Gateway/Request/PaymentDataBuilderTest.php:121]: is_triggered ou triggered

File: Test/Unit/Gateway/Request/PaymentDataBuilderTest.php (hunk 2 above)
Comment (Francois-Gomis) [reply Test/Unit/Gateway/Request/PaymentDataBuilderTest.php:110]: Je ne suis pas sur en français sa donnerai : 
L'option "Trigger" est activé mais le plan n'as pas l'option "trigger"

File: Test/Unit/Gateway/Request/PaymentDataBuilderTest.php (hunk 3 above)
Comment (Francois-Gomis) [reply Test/Unit/Gateway/Request/PaymentDataBuilderTest.php:74]: has_trigger  ?

File: Test/Unit/Gateway/Request/PaymentDataBuilderTest.php (hunk 2 above)
Comment (PaulRouss3l) [reply Test/Unit/Gateway/Request/PaymentDataBuilderTest.php]: on peut faire plus simple alors :)
```
option trigger is on , option plan is off
```

File: Helpers/PaymentValidation.php [hunk 7]
```diff
@@ -393,4 +390,47 @@ public function cancelOrder(Phrase $internalError, bool $transitionOrder, Order
...
+    }
+
+    /**
+     * @param AlmaPayment $almaPayment
+     *
+     * @return bool
+     */
+    private function paymentIsExpired(AlmaPayment $almaPayment): bool
+    {
+        return (bool)$almaPayment->expired_at;
```
Comment (syjust-alma) [inline Helpers/PaymentValidation.php:434]: and what will happen if `expired_at` is after `now()` 😛 (this is not supposed to happen, but could we really be confident with what an API should return us ?)

File: Test/Unit/Gateway/Request/PaymentDataBuilderTest.php [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        $this->locale = $this->createMock(Resolver::class);
+        $this->configHelper = $this->createMock(ConfigHelper::class);
+    }
+
+    /**
+     * @dataProvider paymentPayloadDataProvider
+     *
+     * @return void
+     */
+    public function testNonTriggerPaymentPayload($order, $response): void
```
Comment (syjust-alma) [inline Test/Unit/Gateway/Request/PaymentDataBuilderTest.php:43]:
```suggestion
    public function testTriggerPaymentPayload($order, $response): void
```

Not only for non-trigger 😉

**Review by syjust-alma — APPROVED** [review]
Everything seems OK. Nothing is blocking for the merge.

If you have time and it is not too long, Could you test IPN calls with and without `paymentIsExpired` (by mocking php-client I think) - if you think it's too long to do that, we can keep it for next release, but I think it will be great to add this test.


Next time, please make a PR just for code style / indentation before the feature PR 😛 )

**Pull request merged**